<H1>Vehicles in Finland</H1>

This project is analyzing the registered vehicles in Finland. Dataset is from Trafi https://www.traficom.fi/fi/ajankohtaista/avoin-data?toggle=Ajoneuvojen%20avoin%20data%205.13. Variables are explained in other file Ajoneuvotiedot_avoin_data_luokitukset_28.01.2021.xls


In [1]:
#importing libraries we'll need
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


#dataset is loaded from Trafi's open datasite as zipped csv-file. Using the previous dataset (from 2020) for comparison too.

In [2]:
#read the data
def read_dataset(filename, separ):
    return pd.read_csv(filename, sep=separ, low_memory = False)

raw_data = read_dataset("TieliikenneAvoinData_5_13.csv", ";")

raw_data.tail(5)


ajoneuvoluokka ensirekisterointipvm  ajoneuvoryhma  ajoneuvonkaytto  \
5121142             T1           2020-12-31          929.0              1.0   
5121143             M1           2020-12-31            NaN              1.0   
5121144             O2           2020-12-31           13.0              5.0   
5121145             O2           2020-12-31           13.0              5.0   
5121146             N1           2020-12-31           74.0              1.0   

            variantti                     versio  kayttoonottopvm vari  \
5121142  76505K140F00                        NaN         20201231  NaN   
5121143     VACCPWAX0  FM6FM6AJ009G7MMON1VL74VR2         20150814    9   
5121144           ANP                        270         20201231  NaN   
5121145           NaN                        NaN         20201231  NaN   
5121146      BKRA2ACX              G3MBCSAACAGHA         20190927    9   

         ovienLukumaara korityyppi  ...  vaihteidenLkm  kaupallinenNimi  \
5121142             2.0        NaN  ...            NaN  FENDT 724 VARIO   
5121143             5.0         AC  ...            6.0             GOLF   
5121144             NaN         DC  ...            NaN            C2700   
5121145             NaN         DC  ...            NaN       Eagle 5000   
5121146             5.0         BB  ...            6.0          TRANSIT   

         voimanvalJaTehostamistapa   tyyppihyvaksyntanro  \
5121142                        2.0  e1*167/2013*00273*00   
5121143                        5.0    e1*2007/46*0627*14   
5121144                        8.0    e4*2007/46*0274*06   
5121145                        8.0                   NaN   
5121146                        5.0    e1*2007/46*1100*12   

         yksittaisKayttovoima  kunta   Co2  matkamittarilukema  \
5121142                    02  420.0   NaN                 NaN   
5121143                    13  186.0  94.0            109538.0   
5121144                   NaN  202.0   NaN                 NaN   
5121145                   NaN  846.0   NaN                 NaN   
5121146                    02  179.0   NaN             28298.0   

        valmistenumero2   jarnro  
5121142             NaN  5121143  
5121143      WVWZZZAUZG  5121144  
5121144             NaN  5121145  
5121145             NaN  5121146  
5121146             NaN  5121147  

[5 rows x 38 columns]

#0) käytetään edellistä datasettiä vertailuaineistona rinnalla
raw_data_prev = read_dataset("TieliikenneAvoinData_5_9.csv", ";")
raw_data_prev.tail(5)

In [3]:
#lets see what kind of data we have
print(len(raw_data))
raw_data.describe()

5121147


ajoneuvoryhma  ajoneuvonkaytto  kayttoonottopvm  ovienLukumaara  \
count   2.220690e+06     5.118780e+06     5.121147e+06    2.000290e+06   
mean    1.229783e+02     1.150773e+00     2.000769e+07    4.116740e+00   
std     2.033854e+02     7.673356e-01     8.597438e+05    8.268831e-01   
min     1.000000e+00     1.000000e+00     0.000000e+00    0.000000e+00   
25%     1.000000e+01     1.000000e+00     1.999083e+07    4.000000e+00   
50%     2.100000e+01     1.000000e+00     2.007050e+07    4.000000e+00   
75%     1.090000e+02     1.000000e+00     2.014012e+07    5.000000e+00   
max     9.440000e+02     6.000000e+00     2.021010e+07    5.600000e+01   

       ohjaamotyyppi  istumapaikkojenLkm      omamassa  teknSuurSallKokmassa  \
count  877384.000000        4.002241e+06  5.108321e+06          4.625375e+06   
mean        1.940398        4.130834e+00  1.602662e+03          2.803375e+03   
std         1.381710        2.482750e+00  2.485471e+03          5.373251e+03   
min         1.000000        0.000000e+00  0.000000e+00          0.000000e+00   
25%         1.000000        3.000000e+00  4.700000e+02          1.370000e+03   
50%         1.000000        5.000000e+00  1.345000e+03          1.840000e+03   
75%         4.000000        5.000000e+00  1.671000e+03          2.210000e+03   
max         4.000000        9.100000e+01  9.539560e+05          9.282000e+05   

       tieliikSuurSallKokmassa  ajonKokPituus  ...  iskutilavuus  \
count             3.197434e+06   4.487045e+06  ...  3.993023e+06   
mean              2.833969e+03   4.618558e+03  ...  2.121884e+03   
std               5.716597e+03   1.542978e+03  ...  7.028230e+03   
min               0.000000e+00   1.000000e+01  ...  0.000000e+00   
25%               7.500000e+02   4.146000e+03  ...  1.390000e+03   
50%               1.710000e+03   4.520000e+03  ...  1.798000e+03   
75%               2.160000e+03   4.820000e+03  ...  2.200000e+03   
max               6.800000e+05   4.434000e+05  ...  6.700000e+06   

       suurinNettoteho  sylintereidenLkm  sahkohybridinluokka  vaihteidenLkm  \
count     3.568095e+06      3.529707e+06        121097.000000   1.743278e+06   
mean      9.393813e+01      3.917846e+00             1.622072   5.997244e+00   
std       5.091671e+01      3.761559e+00             0.485059   3.022547e+00   
min       0.000000e+00      0.000000e+00             1.000000   0.000000e+00   
25%       7.100000e+01      4.000000e+00             1.000000   5.000000e+00   
50%       8.800000e+01      4.000000e+00             2.000000   6.000000e+00   
75%       1.100000e+02      4.000000e+00             2.000000   6.000000e+00   
max       1.612000e+04      6.138000e+03             4.000000   3.636000e+03   

       voimanvalJaTehostamistapa         kunta           Co2  \
count               4.167800e+06  5.121090e+06  2.546706e+06   
mean                5.040240e+00  4.342427e+02  1.562626e+02   
std                 2.314670e+00  2.859295e+02  3.998296e+01   
min                 1.000000e+00  5.000000e+00  0.000000e+00   
25%                 5.000000e+00  1.670000e+02  1.270000e+02   
50%                 5.000000e+00  4.100000e+02  1.530000e+02   
75%                 5.000000e+00  6.940000e+02  1.790000e+02   
max                 1.700000e+01  9.990000e+02  2.850000e+03   

       matkamittarilukema        jarnro  
count        2.776312e+06  5.121147e+06  
mean         2.056303e+05  2.560574e+06  
std          1.095331e+06  1.478348e+06  
min         -2.882100e+04  1.000000e+00  
25%          9.676000e+04  1.280288e+06  
50%          1.775290e+05  2.560574e+06  
75%          2.673672e+05  3.840860e+06  
max          1.386690e+09  5.121147e+06  

[8 rows x 22 columns]

In [4]:
cols = raw_data.columns
cols

Index(['ajoneuvoluokka', 'ensirekisterointipvm', 'ajoneuvoryhma',
       'ajoneuvonkaytto', 'variantti', 'versio', 'kayttoonottopvm', 'vari',
       'ovienLukumaara', 'korityyppi', 'ohjaamotyyppi', 'istumapaikkojenLkm',
       'omamassa', 'teknSuurSallKokmassa', 'tieliikSuurSallKokmassa',
       'ajonKokPituus', 'ajonLeveys', 'ajonKorkeus', 'kayttovoima',
       'iskutilavuus', 'suurinNettoteho', 'sylintereidenLkm', 'ahdin',
       'sahkohybridi', 'sahkohybridinluokka', 'merkkiSelvakielinen',
       'mallimerkinta', 'vaihteisto', 'vaihteidenLkm', 'kaupallinenNimi',
       'voimanvalJaTehostamistapa', 'tyyppihyvaksyntanro',
       'yksittaisKayttovoima', 'kunta', 'Co2', 'matkamittarilukema',
       'valmistenumero2', 'jarnro'],
      dtype='object')

In [5]:
#checking how many NaNs in each column
raw_data.isna().sum()

ajoneuvoluokka                     0
ensirekisterointipvm          175203
ajoneuvoryhma                2900457
ajoneuvonkaytto                 2367
variantti                    2273937
versio                       1982070
kayttoonottopvm                    0
vari                         1859631
ovienLukumaara               3120857
korityyppi                   2121177
ohjaamotyyppi                4243763
istumapaikkojenLkm           1118906
omamassa                       12826
teknSuurSallKokmassa          495772
tieliikSuurSallKokmassa      1923713
ajonKokPituus                 634102
ajonLeveys                    341127
ajonKorkeus                  1824203
kayttovoima                  1111894
iskutilavuus                 1128124
suurinNettoteho              1553052
sylintereidenLkm             1591440
ahdin                        2650177
sahkohybridi                 3544306
sahkohybridinluokka          5000050
merkkiSelvakielinen             1797
mallimerkinta                 747146
v

#Note: there are lots of NaNs in ensirekisterointipvm, 175203 -> if ensirekisterointipvm = Nan and kayttoonottopvm Nan or 0, this is problematic when counting stats


In [6]:
#How many registered vehicles we have?
print(f"{len(raw_data)} registered vehicles in Finland after 2020")
#print(f"{len(raw_data_prev)} registered vehicles in Finland after 2019")
#print(f"Difference between 2020 and 2019 is {len(raw_data)-len(raw_data_prev)} vehicles")
raw_data.head()

5121147 registered vehicles in Finland after 2020


ajoneuvoluokka ensirekisterointipvm  ajoneuvoryhma  ajoneuvonkaytto  \
0            MUU                  NaN           21.0              1.0   
1            MUU           1976-09-01           13.0              1.0   
2             M1           1984-07-09            NaN              1.0   
3            MUU           1983-09-22           13.0              1.0   
4             O1           1994-02-09            1.0              1.0   

  variantti versio  kayttoonottopvm vari  ovienLukumaara korityyppi  ...  \
0       NaN    NaN         19670000  NaN             NaN        NaN  ...   
1       NaN    NaN         19760000  NaN             NaN        NaN  ...   
2       NaN    NaN         19840000    1             NaN        NaN  ...   
3       NaN    NaN         19830000  NaN             NaN        NaN  ...   
4       NaN    NaN         19940209  NaN             NaN        NaN  ...   

   vaihteidenLkm  kaupallinenNimi  voimanvalJaTehostamistapa  \
0            NaN              NaN                        NaN   
1            NaN              NaN                        1.0   
2            NaN           FIESTA                        5.0   
3            NaN              NaN                        NaN   
4            NaN              NaN                        NaN   

   tyyppihyvaksyntanro  yksittaisKayttovoima  kunta  Co2  matkamittarilukema  \
0                  NaN                    01   91.0  NaN                 NaN   
1                  NaN                   NaN  475.0  NaN                 NaN   
2                  NaN                    01  179.0  NaN                 NaN   
3                  NaN                   NaN  734.0  NaN                 NaN   
4                  NaN                   NaN  761.0  NaN                 NaN   

  valmistenumero2  jarnro  
0             NaN       1  
1             NaN       2  
2             NaN       3  
3             NaN       4  
4             NaN       5  

[5 rows x 38 columns]

In [7]:
#tracking the year of registration from column ensirekisterointipvm, changing the year to numeric

def pick_the_year(data):
    data["year"] = data["ensirekisterointipvm"].str.slice(stop=4)
    data["kayttoonottopvm"] = (data["kayttoonottopvm"]/10000).map(int)
    data["year"] = data["year"].map(float)
    return data

raw_data=pick_the_year(raw_data)
raw_data.head()


ajoneuvoluokka ensirekisterointipvm  ajoneuvoryhma  ajoneuvonkaytto  \
0            MUU                  NaN           21.0              1.0   
1            MUU           1976-09-01           13.0              1.0   
2             M1           1984-07-09            NaN              1.0   
3            MUU           1983-09-22           13.0              1.0   
4             O1           1994-02-09            1.0              1.0   

  variantti versio  kayttoonottopvm vari  ovienLukumaara korityyppi  ...  \
0       NaN    NaN             1967  NaN             NaN        NaN  ...   
1       NaN    NaN             1976  NaN             NaN        NaN  ...   
2       NaN    NaN             1984    1             NaN        NaN  ...   
3       NaN    NaN             1983  NaN             NaN        NaN  ...   
4       NaN    NaN             1994  NaN             NaN        NaN  ...   

   kaupallinenNimi  voimanvalJaTehostamistapa  tyyppihyvaksyntanro  \
0              NaN                        NaN                  NaN   
1              NaN                        1.0                  NaN   
2           FIESTA                        5.0                  NaN   
3              NaN                        NaN                  NaN   
4              NaN                        NaN                  NaN   

   yksittaisKayttovoima  kunta  Co2  matkamittarilukema  valmistenumero2  \
0                    01   91.0  NaN                 NaN              NaN   
1                   NaN  475.0  NaN                 NaN              NaN   
2                    01  179.0  NaN                 NaN              NaN   
3                   NaN  734.0  NaN                 NaN              NaN   
4                   NaN  761.0  NaN                 NaN              NaN   

  jarnro    year  
0      1     NaN  
1      2  1976.0  
2      3  1984.0  
3      4  1983.0  
4      5  1994.0  

[5 rows x 39 columns]

In [8]:

#there are Nans in kayttoonottopvm, so let's fix this by attaching the ensirekisterointipvm if kayttoonottopvm is nan

def fillnans_with_col(data, colnan, colfill):
    data[colnan] = data[colnan].fillna(data[colfill])
    return data
    
raw_data = fillnans_with_col(raw_data, "year", "kayttoonottopvm")
raw_data.head()

ajoneuvoluokka ensirekisterointipvm  ajoneuvoryhma  ajoneuvonkaytto  \
0            MUU                  NaN           21.0              1.0   
1            MUU           1976-09-01           13.0              1.0   
2             M1           1984-07-09            NaN              1.0   
3            MUU           1983-09-22           13.0              1.0   
4             O1           1994-02-09            1.0              1.0   

  variantti versio  kayttoonottopvm vari  ovienLukumaara korityyppi  ...  \
0       NaN    NaN             1967  NaN             NaN        NaN  ...   
1       NaN    NaN             1976  NaN             NaN        NaN  ...   
2       NaN    NaN             1984    1             NaN        NaN  ...   
3       NaN    NaN             1983  NaN             NaN        NaN  ...   
4       NaN    NaN             1994  NaN             NaN        NaN  ...   

   kaupallinenNimi  voimanvalJaTehostamistapa  tyyppihyvaksyntanro  \
0              NaN                        NaN                  NaN   
1              NaN                        1.0                  NaN   
2           FIESTA                        5.0                  NaN   
3              NaN                        NaN                  NaN   
4              NaN                        NaN                  NaN   

   yksittaisKayttovoima  kunta  Co2  matkamittarilukema  valmistenumero2  \
0                    01   91.0  NaN                 NaN              NaN   
1                   NaN  475.0  NaN                 NaN              NaN   
2                    01  179.0  NaN                 NaN              NaN   
3                   NaN  734.0  NaN                 NaN              NaN   
4                   NaN  761.0  NaN                 NaN              NaN   

  jarnro    year  
0      1  1967.0  
1      2  1976.0  
2      3  1984.0  
3      4  1983.0  
4      5  1994.0  

[5 rows x 39 columns]

In [9]:
raw_data = raw_data[raw_data["kayttoonottopvm"].notna()]
raw_data = raw_data[raw_data["year"].notna()]
len(raw_data) #4945944

5121147

In [10]:
print(f"Nans in raw_data/ensirekisterointipvm {raw_data['ensirekisterointipvm'].isna().sum()}")
print(f"Nans in raw_data/year {raw_data['year'].isna().sum()}")
#dropping rows with Nan or 0 in kayttoonottopvm or year
test = raw_data.apply(lambda x: True if x['year'] == 0 else False , axis=1)
# Count number of True in series
numOfRows = len(test[test == True].index)

Nans in raw_data/ensirekisterointipvm 175203
Nans in raw_data/year 0


In [11]:
print(numOfRows)
test = raw_data.apply(lambda x: True if x['year'] == 0 else False , axis=1)
test2 = raw_data.apply(lambda x: True if (x["year"] == 0) & (x["kayttoonottopvm"]==0) else False, axis=1)

8139


In [12]:
print(len(test2[test2 == True].index))

8139


In [ ]:
test3 = raw_data.apply(lambda x: True if (x["kayttoonottopvm"] == 0) else False, axis=1)

In [ ]:
print(len(test3[test3 == True].index))

In [ ]:
#dropping these 9237 rows
raw_data = raw_data[raw_data.year != 0]
raw_data = raw_data[raw_data.kayttoonottopvm != 0]
len(raw_data) #511910 rows

In [ ]:
print(raw_data["year"].isna().sum(axis=0))
raw_data["year"].describe()
print(((raw_data[raw_data.year < 1.0])).count(), "year < 1")
print(((raw_data[raw_data.kayttoonottopvm < 1.0])).count(), "ko")

In [ ]:
raw_data.head(15)

In [ ]:
# dropping out columns we do not need (variantti,versio,istumapaikkojenLkm, teknSuurSallKokmassa, tieliikSuurSallKokmassa, iskutilavuus, suurinNettoteho, sylintereidenLkm, ahdin, voimanvalJaTehostamistapa, tyyppihyvaksyntanro, yksittaisKayttovoima, valmistenumero2 )
drop_out = ["variantti","versio","istumapaikkojenLkm", "teknSuurSallKokmassa", "tieliikSuurSallKokmassa", "iskutilavuus", "suurinNettoteho", "sylintereidenLkm", "ahdin", "voimanvalJaTehostamistapa", "tyyppihyvaksyntanro", "yksittaisKayttovoima", "valmistenumero2"]

def dropping(data, cols_list):
    data = data.drop(cols_list, axis=1)
    return data

raw_data = dropping(raw_data, drop_out)
raw_data.head()
    


In [ ]:
#how many unique type of vehicles (ajoneuvoluokka) we have?
raw_data["ajoneuvoluokka"].unique()
raw_data["ajoneuvoluokka"].value_counts()

In [ ]:
#Do we have any towed vehicles?
raw_data[raw_data["ajoneuvoluokka"] == "Sb1"]
#zero, keeping this class though for future needs

In [ ]:
#adding now a column with a type of vehicle (more descriptive)
tractor_list = ["C1", "C2", "C3", "C4", "C5", "T", "T1", "T2", "T3", "T4", "T5", "LTR"]
bus_list = ["M2G", "M2", "M3", "M3G"]
lorry_list = ["N2", "N2G", "N3", "N3G"]
van_list = ["N1", "N1G"]
car_list = ["M1", "M1G"]
trailer_list = ["O1", "O2", "O3", "O4", "Ra1", "Ra2", "Ra3", "Ra4", "Rb1", "Rb2", "Rb3", "Rb4"]
motorbike_list = ["L3", "L3e", "L4", "L4e"]
moped_list = ["L1", "L1e", "L2", "L2e"]
light_list = ["KNP", "L5", "L5e", "L6e", "L7e"]
towed_list = ["Sa1", "Sa2", "Sb1", "Sb2"]
other_list = ["MUU", "MTK", "MA"] #MA maastoajoneuvo, MTK moottorityökone

vehicles = {"tractor": tractor_list, "bus": bus_list, "lorry": lorry_list, "van": van_list, "car": car_list, "trailer": trailer_list, "motorbike": motorbike_list, 
           "moped": moped_list, "light": light_list, "towed": towed_list, "other": other_list}

def add_type(data):
    for key, value in vehicles.items():
        data.loc[data["ajoneuvoluokka"].isin(value), "tyyppi"] = key
    

    
    return data

raw_data = add_type(raw_data)
raw_data.head(5)

In [ ]:
raw_data.head()

In [ ]:
#deviding cars to private and professional use cars_priv, cars_pro

def car_priv_pro(data):
    data['tyyppi'] = np.where(
   (data['tyyppi'] == "car") & (data['ajoneuvonkaytto'] == 1.0) , "car_priv", data['tyyppi']
   )
    data['tyyppi'] = np.where(
   (data['tyyppi'] == "car") & (data['ajoneuvonkaytto'] != 1.0) , "car_pro", data['tyyppi']
   )
    return data
    
raw_data = car_priv_pro(raw_data)
raw_data.head(11)

In [ ]:
types = raw_data["tyyppi"].unique()
print(types)
#as we saw earlier, we do not have any towed

In [ ]:
#detecting outliers with seaborn boxplot - these boxplots are not for visualizing dataset, they are for seeing the outliers
sns.boxplot(x=raw_data["kayttoonottopvm"])

#this looks now very reasonable

In [ ]:
sns.boxplot(x=raw_data["year"])

In [ ]:
#any outliers in other columns?
sns.boxplot(x=raw_data["omamassa"])

In [ ]:
#checking what are these two with mass over 125 000 kg
raw_data[raw_data.omamassa >125000]

In [ ]:
sns.boxplot(x=raw_data["matkamittarilukema"])

In [ ]:
raw_data[raw_data.matkamittarilukema > 2000000]
#it's hard to know if these biggest values are wrong or not, with lorries they are ok

In [ ]:
sns.boxplot(x=raw_data["Co2"])

In [ ]:
raw_data[raw_data["Co2"] >1000]
#these are ok

In [ ]:
sns.boxplot(x=raw_data["ajonKokPituus"])

In [ ]:
raw_data[raw_data["ajonKokPituus"] > 150000]
#the unit here is millimeter, so this palju-trailer is an outlier, dropping it out

In [ ]:
print(len(raw_data))
indexRows = raw_data[ (raw_data['ajonKokPituus'] >= 150000)].index
raw_data.drop(indexRows , inplace=True)
print(len(raw_data))

In [ ]:
sns.boxplot(x=raw_data["ajonLeveys"])

In [ ]:
raw_data[raw_data["ajonLeveys"] > 8000]

In [ ]:
#these two motorbikes are mistakes/outliers, the first is a crane, so it is ok -> dropping the motorbikes
indexRows = [2515049, 3148936]
raw_data.drop(indexRows , inplace=True)


In [ ]:
len(raw_data)

In [ ]:
sns.boxplot(x=raw_data["ajonKorkeus"])

In [ ]:
raw_data[raw_data["ajonKorkeus"]>1000000]
#these are outliers, probaply between values combined, dropping them


In [ ]:
raw_data[raw_data["ajonKorkeus"]> 25000]
#this is "normal" tractor, so it is an outlier too

indexRows = raw_data[ (raw_data['ajonKorkeus'] >= 25000)].index
raw_data.drop(indexRows , inplace=True)
len(raw_data)

In [ ]:
sns.boxplot(x=raw_data["ajonKorkeus"])

In [ ]:
#Now data is preprocessed, ready to analyze it, let's save it to csv-file
raw_data.to_csv (r'vehicles_20.csv', header=True)

What kind of vehicles we have? Next phase is to analyze and visualize the dataset.